src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -136,36 +137,39 @@ public async Task<RestoreResult> ExecuteAsync(CancellationToken token)
 
                 CacheFile cacheFile = null;
 
-                using (var noOpTelemetry = TelemetryActivity.Create(parentId: _operationId, eventName: RestoreNoOpInformation))
+                using (telemetry.StartIndependentInterval(NoOpDuration))
                 {
                     if (NoOpRestoreUtilities.IsNoOpSupported(_request))
                     {
-                        noOpTelemetry.StartIntervalMeasure();
-
+                        telemetry.StartIntervalMeasure();
                         bool noOp;
                         (cacheFile, noOp) = EvaluateCacheFile();
-
-                        noOpTelemetry.EndIntervalMeasure(CacheFileEvaluateDuration);
+                        telemetry.TelemetryEvent[NoOpCacheFileEvaluationResult] = noOp;
+                        telemetry.EndIntervalMeasure(NoOpCacheFileEvaluateDuration);
 
                         if (noOp)
                         {
-                            noOpTelemetry.StartIntervalMeasure();
+                            telemetry.StartIntervalMeasure();
 
                             var noOpSuccess = NoOpRestoreUtilities.VerifyRestoreOutput(_request, cacheFile);
 
-                            noOpTelemetry.EndIntervalMeasure(MsbuildAssetsVerificationDuration);
-                            noOpTelemetry.TelemetryEvent[MsbuildAssetsVerificationResult] = noOpSuccess;
+                            telemetry.EndIntervalMeasure(NoOpRestoreOutputEvaluationDuration);
+                            telemetry.TelemetryEvent[NoOpRestoreOutputEvaluationResult] = noOpSuccess;
 
                             if (noOpSuccess)
                             {
-                                noOpTelemetry.StartIntervalMeasure();
+                                telemetry.StartIntervalMeasure();
 
                                 // Replay Warnings and Errors from an existing lock file in case of a no-op.
                                 await MSBuildRestoreUtility.ReplayWarningsAndErrorsAsync(cacheFile.LogMessages, _logger);
 
-                                noOpTelemetry.EndIntervalMeasure(ReplayLogsDuration);
+                                telemetry.EndIntervalMeasure(NoOpReplayLogsDuration);
 
                                 restoreTime.Stop();
+                                telemetry.TelemetryEvent[NoOpResult] = true;
+                                telemetry.TelemetryEvent[RestoreSuccess] = _success;
+                                telemetry.TelemetryEvent[TotalUniquePackagesCount] = cacheFile.ExpectedPackageFilePaths?.Count ?? -1;
```

Comment:
The cache could theoretically be null. 
I didn't want to set 0, because that'd imply there are no packages. but in reality we just couldn't determine. 

By putting `-1` this communicates to us that something is wrong and we need to act on it.

Reply:
I can understand your intention. But I don't see you're making use of `-1` anywhere, can you add unit test?

Reply:
I can't create a scenario like this easily. 
There are other parts in the code that try to be null safe, and this is one of those, `?.`. 

It's unfortunate, and error prone, but I'd be hesitant to remove those safety checks in the context of enhancing telemetry. 

I'd expect that this code is *never* `-1` in actual real world telemetry.